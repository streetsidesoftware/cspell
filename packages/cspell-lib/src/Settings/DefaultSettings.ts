import type {
    CSpellSettings,
    CSpellSettingsWithSourceTrace,
    PredefinedPatterns,
    RegExpPatternDefinition,
} from '@cspell/cspell-types';
import { resolveFile } from '../util/resolveFile';
import { readSettings } from './CSpellSettingsServer';
import { mergeSettings } from './index';
import * as LanguageSettings from './LanguageSettings';
import * as RegPat from './RegExpPatterns';

const defaultConfigFileModuleRef = '@cspell/cspell-bundled-dicts/cspell-default.json';

// Do not use require.resolve because webpack will mess it up.
const defaultConfigFile = resolveConfigModule(defaultConfigFileModuleRef);

const regExpSpellCheckerDisable = [
    RegPat.regExSpellingGuardBlock,
    RegPat.regExSpellingGuardLine,
    RegPat.regExSpellingGuardNext,
];

// cspell:ignore filetypes
const predefinedPatterns = [
    // Exclude patterns
    { name: 'CommitHash', pattern: RegPat.regExCommitHash },
    { name: 'CommitHashLink', pattern: RegPat.regExCommitHashLink },
    { name: 'CStyleHexValue', pattern: RegPat.regExCStyleHexValue },
    { name: 'CSSHexValue', pattern: RegPat.regExCSSHexValue },
    { name: 'Urls', pattern: RegPat.regExMatchUrls },
    { name: 'HexValues', pattern: RegPat.regExMatchCommonHexFormats },
    { name: 'SpellCheckerDisable', pattern: regExpSpellCheckerDisable },
    { name: 'PublicKey', pattern: RegPat.regExPublicKey },
    { name: 'RsaCert', pattern: RegPat.regExCert },
    { name: 'SshRsa', pattern: RegPat.regExSshRSA },
    { name: 'EscapeCharacters', pattern: RegPat.regExEscapeCharacters },
    { name: 'Base64', pattern: RegPat.regExBase64 },
    { name: 'Base64SingleLine', pattern: RegPat.regExBase64SingleLine },
    { name: 'Base64MultiLine', pattern: RegPat.regExBase64MultiLine },
    { name: 'Email', pattern: RegPat.regExEmail },
    { name: 'SHA', pattern: RegPat.regExSha },
    { name: 'HashStrings', pattern: RegPat.regExHashStrings },
    { name: 'UnicodeRef', pattern: RegPat.regExUnicodeRef },
    { name: 'UUID', pattern: RegPat.regExUUID },
    { name: 'href', pattern: RegPat.regExHRef },
    { name: 'SpellCheckerDisableBlock', pattern: RegPat.regExSpellingGuardBlock },
    { name: 'SpellCheckerDisableLine', pattern: RegPat.regExSpellingGuardLine },
    { name: 'SpellCheckerDisableNext', pattern: RegPat.regExSpellingGuardNext },
    { name: 'SpellCheckerIgnoreInDocSetting', pattern: RegPat.regExIgnoreSpellingDirectives },

    // Include Patterns
    { name: 'PhpHereDoc', pattern: RegPat.regExPhpHereDoc },
    { name: 'string', pattern: RegPat.regExString },
    { name: 'CStyleComment', pattern: RegPat.regExCStyleComments },
    { name: 'Everything', pattern: '.*' },
] as const;

type NameType<T> = T extends readonly { name: infer U }[] ? U : never;

type ExtendsType<T, U> = T extends U ? T : never;

type PredefinedPatternNames = ExtendsType<NameType<typeof predefinedPatterns>, PredefinedPatterns>;

const defaultRegExpPatterns: RegExpPatternDefinition[] = [...predefinedPatterns];

const definedDefaultRegExpExcludeList: PredefinedPatterns[] = [
    'SpellCheckerDisable',
    'SpellCheckerIgnoreInDocSetting',
    'Urls',
    'Email',
    'RsaCert',
    'SshRsa',
    'Base64MultiLine',
    'Base64SingleLine',
    'CommitHash',
    'CommitHashLink',
    'CStyleHexValue',
    'CSSHexValue',
    'SHA',
    'HashStrings',
    'UnicodeRef',
    'UUID',
];

// This bit of copying is done to have the complier ensure that the defaults exist.
const defaultRegExpExcludeList: PredefinedPatternNames[] = definedDefaultRegExpExcludeList;

export const _defaultSettings: Readonly<CSpellSettingsWithSourceTrace> = Object.freeze({
    id: 'static_defaults',
    language: 'en',
    name: 'Static Defaults',
    enabled: true,
    enabledLanguageIds: [
        'ada',
        'csharp',
        'go',
        'javascript',
        'javascriptreact',
        'json',
        'markdown',
        'mdx',
        'php',
        'plaintext',
        'python',
        'text',
        'typescript',
        'typescriptreact',
        'haskell',
        'html',
        'css',
        'less',
        'scss',
        'latex',
        'ruby',
        'rust',
        'shellscript',
        'toml',
    ],
    maxNumberOfProblems: 100,
    numSuggestions: 10,
    suggestionsTimeout: 500,
    suggestionNumChanges: 3,
    words: [],
    userWords: [],
    ignorePaths: [],
    allowCompoundWords: false,
    patterns: defaultRegExpPatterns,
    ignoreRegExpList: defaultRegExpExcludeList,
    languageSettings: LanguageSettings.getDefaultLanguageSettings(),
    source: { name: 'defaultSettings' },
    reporters: [],
});

const getSettings = (function () {
    let settings: CSpellSettings | undefined = undefined;
    return function () {
        if (!settings) {
            const jsonSettings = readSettings(defaultConfigFile);
            settings = mergeSettings(_defaultSettings, jsonSettings);
            settings.name = jsonSettings.name;
        }
        return settings;
    };
})();

function resolveConfigModule(configModuleName: string) {
    return resolveFile(configModuleName, __dirname).filename;
}

export function getDefaultSettings(): CSpellSettings {
    return getSettings();
}
