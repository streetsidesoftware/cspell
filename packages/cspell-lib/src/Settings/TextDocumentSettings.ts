import type { CSpellSettings, CSpellUserSettings } from '@cspell/cspell-types';
import { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef';
import * as CSpellSettingsServer from './CSpellSettingsServer';
import { getInDocumentSettings } from './InDocSettings';
import { calcSettingsForLanguageId } from './LanguageSettings';

export function combineTextAndLanguageSettings(
    settings: CSpellUserSettings,
    text: string,
    languageId: string | string[]
): CSpellSettingsInternal {
    const docSettings = extractSettingsFromText(text);
    const settingsForText = CSpellSettingsServer.mergeSettings(settings, docSettings);
    const langSettings = calcSettingsForLanguageId(settingsForText, languageId);
    // Merge again, to force In-Doc settings.
    return CSpellSettingsServer.mergeSettings(langSettings, docSettings);
}

export function extractSettingsFromText(text: string): CSpellSettings {
    return getInDocumentSettings(text);
}
