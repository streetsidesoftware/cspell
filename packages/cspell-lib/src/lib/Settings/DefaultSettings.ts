import type { PredefinedPatterns, RegExpPatternDefinition } from '@cspell/cspell-types';
import { parsers } from 'cspell-grammar';

import { srcDirectory } from '../../lib-cjs/index.cjs';
import type { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { createCSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { PatternRegExp } from '../Models/PatternRegExp.js';
import { resolveFile } from '../util/resolveFile.js';
import { readSettings } from './Controller/configLoader/index.js';
import { mergeSettings } from './CSpellSettingsServer.js';
import * as LanguageSettings from './LanguageSettings.js';
import * as RegPat from './RegExpPatterns.js';

const defaultConfigFileModuleRef = '@cspell/cspell-bundled-dicts/cspell-default.json';

// Do not use require.resolve because webpack will mess it up.
const defaultConfigFile = resolveConfigModule(defaultConfigFileModuleRef);

const regExpSpellCheckerDisable = [
    new PatternRegExp(RegPat.regExSpellingGuardBlock),
    new PatternRegExp(RegPat.regExSpellingGuardLine),
    new PatternRegExp(RegPat.regExSpellingGuardNext),
];

// cspell:ignore filetypes
const predefinedPatterns = [
    // Exclude patterns
    { name: 'CommitHash', pattern: RegPat.regExCommitHash },
    { name: 'CommitHashLink', pattern: RegPat.regExCommitHashLink },
    { name: 'CStyleHexValue', pattern: RegPat.regExCStyleHexValue },
    { name: 'CSSHexValue', pattern: RegPat.regExCSSHexValue },
    { name: 'Urls', pattern: RegPat.regExMatchUrls },
    { name: 'HexValues', pattern: RegPat.regExMatchCommonHexFormats },
    { name: 'SpellCheckerDisable', pattern: regExpSpellCheckerDisable },
    { name: 'PublicKey', pattern: RegPat.regExPublicKey },
    { name: 'RsaCert', pattern: RegPat.regExCert },
    { name: 'SshRsa', pattern: RegPat.regExSshRSA },
    { name: 'EscapeCharacters', pattern: RegPat.regExEscapeCharacters },
    { name: 'Base64', pattern: RegPat.regExBase64 },
    { name: 'Base64SingleLine', pattern: RegPat.regExBase64SingleLine },
    { name: 'Base64MultiLine', pattern: RegPat.regExBase64MultiLine },
    { name: 'Email', pattern: RegPat.regExEmail },
    { name: 'SHA', pattern: RegPat.regExSha },
    { name: 'HashStrings', pattern: RegPat.regExHashStrings },
    { name: 'UnicodeRef', pattern: RegPat.regExUnicodeRef },
    { name: 'UUID', pattern: RegPat.regExUUID },
    { name: 'href', pattern: RegPat.regExHRef },
    { name: 'SpellCheckerDisableBlock', pattern: RegPat.regExSpellingGuardBlock },
    { name: 'SpellCheckerDisableLine', pattern: RegPat.regExSpellingGuardLine },
    { name: 'SpellCheckerDisableNext', pattern: RegPat.regExSpellingGuardNext },
    { name: 'SpellCheckerIgnoreInDocSetting', pattern: RegPat.regExIgnoreSpellingDirectives },

    // Include Patterns
    { name: 'PhpHereDoc', pattern: RegPat.regExPhpHereDoc },
    { name: 'string', pattern: RegPat.regExString },
    { name: 'CStyleComment', pattern: RegPat.regExCStyleComments },
    { name: 'Everything', pattern: '.*' },
] as const;

type NameType<T> = T extends readonly { name: infer U }[] ? U : never;

type ExtendsType<T, U> = T extends U ? T : never;

type PredefinedPatternNames = ExtendsType<NameType<typeof predefinedPatterns>, PredefinedPatterns>;

const defaultRegExpPatterns: RegExpPatternDefinition[] = [...predefinedPatterns].map(normalizePattern);

const definedDefaultRegExpExcludeList: PredefinedPatterns[] = [
    'SpellCheckerDisable',
    'SpellCheckerIgnoreInDocSetting',
    'Urls',
    'Email',
    'RsaCert',
    'SshRsa',
    'Base64MultiLine',
    'Base64SingleLine',
    'CommitHash',
    'CommitHashLink',
    'CStyleHexValue',
    'CSSHexValue',
    'SHA',
    'HashStrings',
    'UnicodeRef',
    'UUID',
];

// This bit of copying is done to have the complier ensure that the defaults exist.
const defaultRegExpExcludeList: PredefinedPatternNames[] = definedDefaultRegExpExcludeList;

export const _defaultSettingsBasis: Readonly<CSpellSettingsInternal> = Object.freeze(
    createCSpellSettingsInternal({
        id: 'static_defaults',
        language: 'en',
        name: 'Static Defaults',
        enabled: true,
        enabledLanguageIds: [],
        maxNumberOfProblems: 100,
        numSuggestions: 10,
        suggestionsTimeout: 500,
        suggestionNumChanges: 3,
        words: [],
        userWords: [],
        ignorePaths: [],
        allowCompoundWords: false,
        patterns: defaultRegExpPatterns,
        ignoreRegExpList: [],
        languageSettings: [],
        source: { name: 'defaultSettings' },
        reporters: [],
        plugins: [{ parsers }],
    }),
);

export const _defaultSettings: Readonly<CSpellSettingsInternal> = Object.freeze(
    createCSpellSettingsInternal({
        ..._defaultSettingsBasis,
        enabledLanguageIds: [
            'ada',
            'csharp',
            'go',
            'javascript',
            'javascriptreact',
            'json',
            'markdown',
            'mdx',
            'php',
            'plaintext',
            'python',
            'text',
            'typescript',
            'typescriptreact',
            'haskell',
            'html',
            'css',
            'less',
            'scss',
            'latex',
            'ruby',
            'rust',
            'shellscript',
            'toml',
        ],
        ignoreRegExpList: defaultRegExpExcludeList,
        languageSettings: LanguageSettings.getDefaultLanguageSettings(),
    }),
);

const getSettings = (function () {
    let settings: CSpellSettingsInternal | undefined = undefined;
    return function (useDefaultDictionaries: boolean) {
        if (!useDefaultDictionaries) {
            return _defaultSettingsBasis;
        }
        if (!settings) {
            const jsonSettings = readSettings(defaultConfigFile);
            settings = mergeSettings(_defaultSettings, jsonSettings);
            if (jsonSettings.name !== undefined) {
                settings.name = jsonSettings.name;
            } else {
                delete settings.name;
            }
        }
        return settings;
    };
})();

function resolveConfigModule(configModuleName: string) {
    return resolveFile(configModuleName, srcDirectory).filename;
}

function normalizePattern(pat: RegExpPatternDefinition): RegExpPatternDefinition {
    const { name, pattern, description } = pat;
    if (!(pattern instanceof RegExp)) return pat;

    return {
        name,
        pattern: new PatternRegExp(pattern),
        description,
    };
}

export function getDefaultSettings(useDefaultDictionaries = true): CSpellSettingsInternal {
    return getSettings(useDefaultDictionaries);
}

export function getDefaultBundledSettings(): CSpellSettingsInternal {
    return getDefaultSettings();
}
