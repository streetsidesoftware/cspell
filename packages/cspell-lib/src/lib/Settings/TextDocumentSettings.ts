import type { CSpellSettings, CSpellUserSettings } from '@cspell/cspell-types';

import type { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { mergeSettings, toInternalSettings } from './CSpellSettingsServer.js';
import { getInDocumentSettings } from './InDocSettings.js';
import { calcSettingsForLanguageId } from './LanguageSettings.js';

export function combineTextAndLanguageSettings(
    settings: CSpellUserSettings,
    text: string | undefined,
    languageId: string | string[],
): CSpellSettingsInternal {
    if (!text) {
        return toInternalSettings(calcSettingsForLanguageId(settings, languageId));
    }
    const docSettings = extractSettingsFromText(text);
    const settingsForText = mergeSettings(settings, docSettings);
    const langSettings = calcSettingsForLanguageId(settingsForText, languageId);
    // Merge again, to force In-Doc settings.
    return mergeSettings(langSettings, docSettings);
}

export function extractSettingsFromText(text: string): CSpellSettings {
    return getInDocumentSettings(text);
}
