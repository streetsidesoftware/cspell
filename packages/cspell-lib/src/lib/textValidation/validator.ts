import type { CSpellUserSettings } from '@cspell/cspell-types';
import { IssueType } from '@cspell/cspell-types';

import { createTextDocument } from '../Models/TextDocument.js';
import type { ValidationIssue } from '../Models/ValidationIssue.js';
import * as Settings from '../Settings/index.js';
import type { DirectiveIssue } from '../Settings/InDocSettings.js';
import { validateInDocumentSettings } from '../Settings/InDocSettings.js';
import { CompoundWordsMethod, getDictionaryInternal } from '../SpellingDictionary/index.js';
import { settingsToValidateOptions } from './settingsToValidateOptions.js';
import { validateText as validateFullText } from './textValidator.js';
import type { ValidateTextOptions } from './ValidateTextOptions.js';

export const diagSource = 'cSpell Checker';

/**
 * @deprecated
 * @deprecationMessage Use spellCheckDocument
 */
export async function validateText(
    text: string,
    settings: CSpellUserSettings,
    options: ValidateTextOptions = {}
): Promise<ValidationIssue[]> {
    const finalSettings = Settings.finalizeSettings(settings);
    const dict = await getDictionaryInternal(finalSettings);
    const spellingIssues = [...validateFullText(text, dict, settingsToValidateOptions(finalSettings))];
    const validationIssues =
        options.validateDirectives || finalSettings.validateDirectives
            ? validateInDocumentSettings(text, settings)
            : [];
    const issues: ValidationIssue[] = spellingIssues.concat(mapValidationIssues(text, validationIssues));
    if (!options.generateSuggestions) {
        return issues;
    }
    const sugOptions = {
        numSuggestions: options.numSuggestions,
        compoundMethod: CompoundWordsMethod.NONE,
        includeTies: false,
        ignoreCase: !(settings.caseSensitive ?? false),
        timeout: settings.suggestionsTimeout,
        numChanges: settings.suggestionNumChanges,
    };
    const withSugs = issues.map((t) => {
        const text = t.text;
        const suggestionsEx = dict
            .suggest(text, sugOptions)
            .map(({ word, isPreferred }) => (isPreferred ? { word, isPreferred } : { word }));
        t.suggestions = suggestionsEx.map((s) => s.word);
        t.suggestionsEx = suggestionsEx;
        return t;
    });

    return withSugs;
}

function mapValidationIssues(text: string, valIssues: Iterable<DirectiveIssue>): ValidationIssue[] {
    const issues = [...valIssues];
    if (!issues.length) return [];

    const document = createTextDocument({ uri: '', content: text });
    const issueType = IssueType.directive;

    function toValidationIssue(dirIssue: DirectiveIssue): ValidationIssue {
        const { text, range, suggestions, suggestionsEx, message } = dirIssue;
        const offset = range[0];
        const pos = document.positionAt(offset);
        const line = document.getLine(pos.line);
        const issue: ValidationIssue = { text, offset, line, suggestions, suggestionsEx, message, issueType };

        return issue;
    }

    return issues.map(toValidationIssue);
}
