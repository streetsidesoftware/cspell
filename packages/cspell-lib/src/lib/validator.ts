export type {
    CheckTextInfo,
    IncludeExcludeOptions,
    TextInfoItem,
    ValidateTextOptions,
    ValidationIssue,
} from './textValidation/index.js';
export { checkText, checkTextDocument, IncludeExcludeFlag, validateText } from './textValidation/index.js';
