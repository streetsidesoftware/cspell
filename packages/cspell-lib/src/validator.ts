export { checkText, checkTextDocument, IncludeExcludeFlag, validateText } from './textValidation';
export type {
    CheckTextInfo,
    IncludeExcludeOptions,
    TextInfoItem,
    ValidateTextOptions,
    ValidationIssue,
} from './textValidation';
