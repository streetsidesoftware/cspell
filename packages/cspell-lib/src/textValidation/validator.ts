import type { CSpellUserSettings } from '@cspell/cspell-types';
import * as Settings from '../Settings';
import { CompoundWordsMethod, getDictionaryInternal } from '../SpellingDictionary';
import { clean } from '../util/util';
import type { ValidationOptions, ValidationResult } from './textValidator';
import { calcTextInclusionRanges, validateText as validateFullText } from './textValidator';

export const diagSource = 'cSpell Checker';

export interface ValidationIssue extends ValidationResult {
    suggestions?: string[];
}

export interface ValidateTextOptions {
    /** Generate suggestions where there are spelling issues. */
    generateSuggestions?: boolean;
    /** The number of suggestions to generate. The higher the number the longer it takes. */
    numSuggestions?: number;
}

export async function validateText(
    text: string,
    settings: CSpellUserSettings,
    options: ValidateTextOptions = {}
): Promise<ValidationIssue[]> {
    const finalSettings = Settings.finalizeSettings(settings);
    const dict = await getDictionaryInternal(finalSettings);
    const issues = [...validateFullText(text, dict, settingsToValidateOptions(finalSettings))];
    if (!options.generateSuggestions) {
        return issues;
    }
    const withSugs = issues.map((t) => {
        const suggestions = dict
            .suggest(
                t.text,
                clean({
                    numSuggestions: options.numSuggestions,
                    compoundMethod: CompoundWordsMethod.NONE,
                    includeTies: false,
                    ignoreCase: !(settings.caseSensitive ?? false),
                    timeout: settings.suggestionsTimeout,
                    numChanges: settings.suggestionNumChanges,
                })
            )
            .map((r) => r.word);
        return { ...t, suggestions };
    });

    return withSugs;
}

export function settingsToValidateOptions(settings: CSpellUserSettings): ValidationOptions {
    const opt: ValidationOptions = {
        ...settings,
        ignoreCase: !(settings.caseSensitive ?? false),
    };
    return opt;
}

export interface CheckTextInfo {
    // Full text
    text: string;
    // Set of include items
    items: TextInfoItem[];
}

export interface TextInfoItem {
    // the segment of text that is either include or excluded
    text: string;
    startPos: number;
    endPos: number;
    flagIE: IncludeExcludeFlag;
    isError?: boolean;
}

export enum IncludeExcludeFlag {
    INCLUDE = 'I',
    EXCLUDE = 'E',
}

export async function checkText(text: string, settings: CSpellUserSettings): Promise<CheckTextInfo> {
    const validationResult = validateText(text, settings);
    const finalSettings = Settings.finalizeSettings(settings);
    const includeRanges = calcTextInclusionRanges(text, finalSettings);
    const result: TextInfoItem[] = [];
    let lastPos = 0;
    for (const { startPos, endPos } of includeRanges) {
        result.push({
            text: text.slice(lastPos, startPos),
            startPos: lastPos,
            endPos: startPos,
            flagIE: IncludeExcludeFlag.EXCLUDE,
        });
        result.push({
            text: text.slice(startPos, endPos),
            startPos,
            endPos,
            flagIE: IncludeExcludeFlag.INCLUDE,
        });
        lastPos = endPos;
    }
    result.push({
        text: text.slice(lastPos),
        startPos: lastPos,
        endPos: text.length,
        flagIE: IncludeExcludeFlag.EXCLUDE,
    });

    const issues = await validationResult;

    function* merge() {
        let i = 0;
        for (const r of result) {
            if (i >= issues.length || issues[i].offset >= r.endPos) {
                yield r;
                continue;
            }
            const span = { ...r };
            while (i < issues.length && issues[i].offset < span.endPos) {
                const issue = issues[i];
                const endPos = issue.offset;
                const text = span.text.slice(0, endPos - span.startPos);
                const endPosError = issue.offset + issue.text.length;
                yield { ...span, text, endPos };
                yield {
                    ...span,
                    isError: true,
                    startPos: issue.offset,
                    endPos: endPosError,
                    text: issue.text,
                };
                span.text = span.text.slice(endPosError - span.startPos);
                span.startPos = endPosError;
                i += 1;
            }
            yield span;
        }
    }

    return {
        text,
        items: [...merge()].filter((i) => i.startPos < i.endPos),
    };
}
