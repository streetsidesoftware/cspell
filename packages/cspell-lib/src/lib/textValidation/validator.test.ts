import type { CSpellSettings } from '@cspell/cspell-types';
import { loremIpsum } from 'lorem-ipsum';
import { describe, expect, test } from 'vitest';

import { getDefaultSettings } from '../Settings/DefaultSettings.js';
import { mergeSettings } from '../Settings/index.js';
import * as tds from '../Settings/TextDocumentSettings.js';
import type { ValidationIssue } from '../validator.js';
import * as Validator from './validator.js';

// cSpell:ignore brouwn jumpped lazzy wrongg mispelled ctrip nmove mischecked

const ac = expect.arrayContaining;
const notAc = expect.not.arrayContaining;

const oc = expect.objectContaining;

describe('Validator', () => {
    test('validates the validator', async () => {
        const text = 'The quick brouwn fox jumpped over the lazzy dog.';
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words).toEqual(['brouwn', 'jumpped', 'lazzy']);
    });

    test('validates ignore Case', async () => {
        const text = 'The Quick brown fox Jumped over the lazy dog.';
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words).toEqual([]);
    });

    test('validate limit', async () => {
        const text = loremIpsum({ count: 5, units: 'paragraphs' });
        const languageId = 'plaintext';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        expect(results).toHaveLength(10);
    });

    test('validates reserved words', async () => {
        const text = 'constructor const prototype type typeof null undefined';
        const languageId = 'javascript';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        expect(results).toHaveLength(0);
    });

    test('validates regex inclusions/exclusions', async () => {
        const text = sampleCode;
        const languageId = 'plaintext';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.not.arrayContaining(['xaccd']));
        expect(words).toEqual(expect.not.arrayContaining(['ctrip']));
        expect(words).toEqual(expect.not.arrayContaining(['FFEE']));
        expect(words).toEqual(expect.not.arrayContaining(['nmove']));
    });

    test('validates ignoreRegExpList', async () => {
        const text = sampleCode;
        const languageId = 'plaintext';
        const settings = {
            ...getSettings(text, languageId),
            maxNumberOfProblems: 10,
            ignoreRegExpList: ['^const [wy]RON[g]+', 'mis.*led'],
        };
        const results = await Validator.validateText(text, settings);
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.not.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.not.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.arrayContaining(['mischecked']));
    });

    test('validates ignoreRegExpList 2', async () => {
        const results = await Validator.validateText(sampleCode, {
            ignoreRegExpList: ['/^const [wy]ron[g]+/gim', '/MIS...LED/g', '/mischecked'],
        });
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.not.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.arrayContaining(['mischecked']));
    });

    test.each`
        settings                                             | expected                        | message
        ${{ ignoreRegExpList: ['/wrong[/gim', 'mis.*led'] }} | ${ac(['wrongg', 'mischecked'])} | ${'malformed ignoreRegExpList'}
        ${{ ignoreRegExpList: ['/wrong[/gim', 'mis.*led'] }} | ${notAc(['mispelled'])}         | ${'malformed ignoreRegExpList'}
        ${{}}                                                | ${notAc(['worlds'])}            | ${''}
        ${{ validateDirectives: true }}                      | ${ac(['worlds'])}               | ${'Invalid directive'}
    `('validates $message', async ({ settings, expected }) => {
        const langSettings = getSettings(sampleCode, 'plaintext');
        const finalSettings = mergeSettings(langSettings, settings);
        const results = await Validator.validateText(sampleCode, finalSettings);
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expected);
    });

    // cspell:ignore hellosd applesq bananasa respectss
    test('Issue #7', async () => {
        const text = `Fails to detect obviously misspelt words, such as:
            hellosd
            applesq
            bananasa
            respectss
        `;
        const expected = ['hellosd', 'applesq', 'bananasa', 'respectss'];
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words.sort()).toEqual(expected.sort());
    });

    test('Validates contractions', async () => {
        const text = `
            We have a bit of text to check. Don't look too hard.
            Which single quote to use? Is it shouldn't or shouldnâ€™t?
        `;
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words.sort()).toEqual([]);
    });

    // cspell:ignore witth feww mistaks
    test('validateText with suggestions', async () => {
        const text = `
            Here is a bit of text witth a feww mistaks.
        `;
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const result = await Validator.validateText(text, settings, { generateSuggestions: true, numSuggestions: 5 });
        expect(result).toMatchSnapshot();
    });

    // const isFoundTrue = { isFound: true };
    const isFoundFalse = { isFound: false };
    const isFlaggedTrue = { isFlagged: true };
    const isFlaggedFalse = { isFlagged: false };

    // cspell:ignore grappes colour
    test.each`
        text              | expected
        ${'hello'}        | ${[]}
        ${'flagged'}      | ${[mValIssue('flagged', isFlaggedTrue)]}
        ${'grappes'}      | ${[mValIssue('grappes', isFoundFalse, isFlaggedFalse)]}
        ${'colour'}       | ${[mValIssue('colour', isFlaggedTrue)]}
        ${'ignored'}      | ${[]}
        ${'crazzzy-code'} | ${[]}
        ${'hyphen-wordz'} | ${[]}
    `('validation "$text"', async ({ text, expected }) => {
        const settings = sampleSettings();
        const languageId = 'plaintext';
        const docSettings = tds.combineTextAndLanguageSettings(settings, text, languageId);
        const results = await Validator.validateText(text, docSettings);
        expect(results).toEqual(expected);
    });
});

// cspell:ignore xaccd ffee

const sampleCode = `

// Verify urls do not get checked.
const url = 'http://ctrip.com?q=words';

// Verify hex values.
const value = 0xaccd;

/* spell-checker:disable */

const weirdWords = ['ctrip', 'xebia', 'zando', 'zooloo'];

/* spell-checker:enable */

// spell:worlds

const wrongg = 'mispelled';
const check = 'mischecked';
const message = "\\nmove to next line";

const hex = 0xBADC0FFEE;

`;

const sampleWords = [
    'and',
    'ant',
    'apple',
    'ate',
    'big',
    'elephant',
    'giraffe',
    'grape',
    'little',
    'mango',
    'orange',
    'purple',
    'the',
    'tiger',
    'worm',
];

// cspell:ignore hte
const flagWords = ['hte', 'flagged', 'ignored'];
// cspell:ignore behaviour
const rejectWords = ['!colour', '!behaviour', '!favour'];

// cspell:ignore crazzzy wordz
const ignoreWords = ['ignored', 'crazzzy-code'];

const words = sampleWords.concat(rejectWords).concat(['hyphen-wordz']);

const _defaultSettings = await getDefaultSettings(true);

const sampleCSpell: CSpellSettings = {
    ..._defaultSettings,
    version: '0.2',
    flagWords,
    words,
    ignoreWords,
};

const defaultSettings: CSpellSettings = {
    ..._defaultSettings,
    enabledLanguageIds: ['plaintext', 'javascript'],
};

function mValIssue(text: string, ...parts: Partial<ValidationIssue>[]): ValidationIssue {
    const issue: Partial<ValidationIssue> = {
        text,
    };
    for (const p of parts) {
        Object.assign(issue, p);
    }
    return oc(issue);
}

function sampleSettings() {
    return Object.freeze({ ...sampleCSpell });
}

function getSettings(text: string, languageId: string) {
    return tds.combineTextAndLanguageSettings(defaultSettings, text, languageId);
}
