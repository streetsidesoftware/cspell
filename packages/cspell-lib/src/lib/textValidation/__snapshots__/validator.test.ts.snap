// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validator > validateText with suggestions 1`] = `
[
  {
    "isFlagged": false,
    "isFound": false,
    "line": {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 35,
    "suggestions": [
      "with",
      "witt",
      "witty",
      "Witt",
      "witch",
    ],
    "suggestionsEx": [
      {
        "word": "with",
      },
      {
        "word": "witt",
      },
      {
        "word": "witty",
      },
      {
        "word": "Witt",
      },
      {
        "word": "witch",
      },
    ],
    "text": "witth",
  },
  {
    "isFlagged": false,
    "isFound": false,
    "line": {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 43,
    "suggestions": [
      "few",
      "flew",
      "faw",
      "dew",
      "hew",
    ],
    "suggestionsEx": [
      {
        "word": "few",
      },
      {
        "word": "flew",
      },
      {
        "word": "faw",
      },
      {
        "word": "dew",
      },
      {
        "word": "hew",
      },
    ],
    "text": "feww",
  },
  {
    "isFlagged": false,
    "isFound": false,
    "line": {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 48,
    "suggestions": [
      "mistake",
      "mistakes",
      "mistaken",
      "mistaker",
      "mistake's",
    ],
    "suggestionsEx": [
      {
        "word": "mistake",
      },
      {
        "word": "mistakes",
      },
      {
        "word": "mistaken",
      },
      {
        "word": "mistaker",
      },
      {
        "word": "mistake's",
      },
    ],
    "text": "mistaks",
  },
]
`;
